plugins/compositing_gl/src/AO.cpp

```diff
@@ -0,0 +1,392 @@
+    this->lighting_prgm_->setUniform("inDensityTex", static_cast<int>(3));
+
+
+    this->lighting_prgm_->setUniform("inWidth", static_cast<float>(cur_vp_width_));
+    this->lighting_prgm_->setUniform("inHeight", static_cast<float>(cur_vp_height_));
+    glUniformMatrix4fv(
+        this->lighting_prgm_->getUniformLocation("MVPinv"), 1, GL_FALSE, glm::value_ptr(this->cur_mvp_inv_));
+    if (enable_lighting) {
+        this->lighting_prgm_->setUniform("inObjLightDir", glm::vec3(cur_light_dir_));
+        this->lighting_prgm_->setUniform("inObjCamPos", cur_cam_pos_);
+    }
+    this->lighting_prgm_->setUniform("inAOOffset", this->ao_offset_slot_.Param<param::FloatParam>()->Value());
+    this->lighting_prgm_->setUniform("inAOStrength", this->ao_strength_slot_.Param<param::FloatParam>()->Value());
+    this->lighting_prgm_->setUniform("inAOConeLength", this->ao_cone_length_slot_.Param<param::FloatParam>()->Value());
+
+    // calculate amb cone constants
+    vislib::math::Dimension<float, 3> dims = this->cur_clip_box_.GetSize();
+    // Calculate the extensions of the volume by using the specified number of voxels for the longest edge
+    float longest_edge = this->cur_clip_box_.LongestEdge();
+    dims.Scale(static_cast<float>(vol_size) / longest_edge);
+    // The X size must be a multiple of 4, so we might have to correct that a little
+    dims.SetWidth(ceil(dims.GetWidth() / 4.0f) * 4.0f);
+    dims.SetHeight(ceil(dims.GetHeight()));
+    dims.SetDepth(ceil(dims.GetDepth()));
+    float amb_cone_constant_0 = std::min(dims.Width(), std::min(dims.Height(), dims.Depth()));
+    float amb_cone_constant_1 = ceil(std::log2(static_cast<float>(vol_size))) - 1.0f;
+
+    this->lighting_prgm_->setUniform("inAmbVolShortestEdge", amb_cone_constant_0);
+    this->lighting_prgm_->setUniform("inAmbVolMaxLod", amb_cone_constant_1);
+    glm::vec3 cur_clip_box_coords = glm::vec3(this->cur_clip_box_.GetLeftBottomBack().GetX(),
+        this->cur_clip_box_.GetLeftBottomBack().GetY(), this->cur_clip_box_.GetLeftBottomBack().GetZ());
+    this->lighting_prgm_->setUniform("inBoundsMin", cur_clip_box_coords);
+    glm::vec3 cur_clip_box_size = glm::vec3(this->cur_clip_box_.GetSize().GetWidth(),
+        this->cur_clip_box_.GetSize().GetHeight(), this->cur_clip_box_.GetSize().GetDepth());
+    this->lighting_prgm_->setUniform("inBoundsSize", cur_clip_box_size);
+
+    // Draw screen filling 'quad' (2 triangle, front facing: CCW)
+    std::vector<GLfloat> vertices = {-1.0f, -1.0f, 1.0f, 1.0f, -1.0f, 1.0f, -1.0f, -1.0f, 1.0f, -1.0f, 1.0f, 1.0f};
+    GLuint vert_attrib_loc = glGetAttribLocation(this->lighting_prgm_->getHandle(), "inPosition");
+    glEnableVertexAttribArray(vert_attrib_loc);
+    glVertexAttribPointer(vert_attrib_loc, 2, GL_FLOAT, GL_TRUE, 0, vertices.data());
+    glDrawArrays(GL_TRIANGLES, static_cast<GLint>(0), static_cast<GLsizei>(vertices.size() / 2));
+    glDisableVertexAttribArray(vert_attrib_loc);
```

Comment:
It should suffice to do the quad preparation once before rendering, instead of every frame.